### PR TITLE
Borůvka's algorithm, spanning tree algorithms bundle and interfaces

### PR DESCRIPTION
**Pull request by d-michail**
Main changes
 - Added Borůvka's algorithm for the computation of a minimum spanning tree
 - Revised spanning tree and spanner interfaces
 - Bundled spanning tree and spanner algorithms into package org.jgrapht.alg.spanning
 - Added tests for Borůvka's algorithm
 - Added tests which construct random graphs and check that all spanning tree algorithms return the same weight
 - Added benchmark which executes all minimum spanning tree algorithms on random graphs.
 - Small improvements in `StopWatch` class

### CODE REVIEW

**Comment by jkinable** [comment]
Great work! I'm currently traveling for a conference. I'll be back and finished with my deadlines on Nov 22. I'll allocate a week to clear my reviewing backlog :)

**Comment by d-michail** [comment]
No problem, I will also find the time to change the interfaces to accept the graph instance in the constructor, similar to the shortest path algorithms!

File: jgrapht-core/src/test/java/org/jgrapht/perf/spanning/MinimumSpanningTreePerformanceTest.java [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+    {
+        System.out.println("Minimum Spanning Tree Benchmark");
+        System.out.println("-------------------------------");
+        System.out.println(
+            "Using G(n,p) random graph with n = " + PERF_BENCHMARK_VERTICES_COUNT + ", p = "
+                + PERF_BENCHMARK_EDGES_PROP);
+        System.out.println("Warmup phase " + WARMUP_REPEAT + " executions");
+        System.out.println("Averaging results over " + REPEAT + " executions");
+
+        List<Supplier<BenchmarkBase>> algFactory = new ArrayList<>();
```
Comment (jkinable) [inline jgrapht-core/src/test/java/org/jgrapht/perf/spanning/MinimumSpanningTreePerformanceTest.java]: Out of curiosity, what's the advantage of this supplier/factory approach versus:
BenchmarkBase[] benchmarks={new PrimBenchmark(), new KruskalBenchmark(), new BoruvkaBenchmark()};
for(BenchmarkBase benchmark: benchmarks){...}

File: jgrapht-core/src/test/java/org/jgrapht/perf/spanning/MinimumSpanningTreePerformanceTest.java [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+        }
+
+        @Override
+        public String toString()
+        {
+            return "Boruvka";
+        }
+    }
+
+    public void testBenchmark()
```
Comment (jkinable) [inline jgrapht-core/src/test/java/org/jgrapht/perf/spanning/MinimumSpanningTreePerformanceTest.java]: Is there a reason why you used the Stopwatch class instead of the build-in mechanisms of jmh like you did for the PathGrowingWeightedMatchingPerformanceTest class? Personally I would say this approach is fine, but I'm not an expert in building proper performance tests.

File: jgrapht-core/src/test/java/org/jgrapht/perf/spanning/MinimumSpanningTreePerformanceTest.java (hunk 1 above)
Comment (d-michail) [inline jgrapht-core/src/test/java/org/jgrapht/perf/spanning/MinimumSpanningTreePerformanceTest.java]: I used to have an outer loop which also switched graph implementations (one with fast lookup specifics and one without), and performed all combinations (graph implementation x algorithm). 

With the factory, it is easier to plug it in somewhere where the tests need to be run multiple times depending of some other parameter changing.

File: jgrapht-core/src/test/java/org/jgrapht/perf/spanning/MinimumSpanningTreePerformanceTest.java (hunk 2 above)
Comment (d-michail) [inline jgrapht-core/src/test/java/org/jgrapht/perf/spanning/MinimumSpanningTreePerformanceTest.java]: While the jmh is really nice, it seems to be tailored more towards systems engineering and not algorithms engineering. In the previous benchmarks it felt somehow that I was abusing it to get the functionality that was required. 

With respect to the approximate matching algorithms, I tried to compare the  approximation factor of the algorithms and did not really succeed. They have something called [AuxCounters](http://hg.openjdk.java.net/code-tools/jmh/file/2e94f4ff263b/jmh-samples/src/main/java/org/openjdk/jmh/samples/JMHSample_23_AuxCounters.java) but they are really limited, and they even warn against using them in the javadoc (they call it an experimental api, which may be removed in the future, etc.).

Moreover, I seem to be having trouble executing the tests after serious changes (something about some manifest file missing). This is probably my fault, but still the current setup is a bit difficult to use.

The approach that I used in this case, is the standard way of comparing algorithms. Some people like to also test with a cold cache which can be done by setting the `WARMUP_REPEAT` parameter to zero (and of course clear the system's memory buffer cache, using for example /proc/sys/vm/drop_caches in linux).

**Comment by d-michail** [comment]
Updated the interfaces to conform with be consistent with the ones for flows and matchings.

**Comment by jkinable** [comment]
Alright, yet another great addition to the library. Thanks!
